ibverbs/src/lib.rs

```diff
@@ -533,6 +552,94 @@ impl<'ctx> CompletionQueue<'ctx> {
             Ok(&mut completions[0..n as usize])
         }
     }
+
+    /// Waits for one or more work completions in a Completion Queue (CQ).
+    ///
+    /// Unlike `poll`, this method blocks until at least one work completion is available or the
+    /// optional timeout expires. It is designed to wait efficiently for completions when polling
+    /// alone is insufficient, such as in low-traffic scenarios.
+    ///
+    /// The returned slice reflects completed work requests (e.g., sends, receives) from the
+    /// associated Work Queue. Not all fields in `ibv_wc` are valid unless the status is
+    /// `IBV_WC_SUCCESS`.
+    ///
+    /// # Errors
+    /// - `TimedOut`: If the timeout expires before any completions are available.
+    /// - System errors: From underlying calls like `req_notify_cq`, `poll`, or `ibv_get_cq_event`.
+    ///
+    /// # Notes
+    /// - This method blocks, unlike `poll`, but avoids busy-waiting by leveraging CQ events and
+    ///   system polling.
+    /// - Callers must ensure the CQ does not overrun (exceed its capacity), as this triggers an
+    ///   `IBV_EVENT_CQ_ERR` async event, rendering the CQ unusable.
+    pub fn wait<'c>(
+        &self,
+        completions: &'c mut [ffi::ibv_wc],
+        timeout: Option<Duration>,
+    ) -> io::Result<&'c mut [ffi::ibv_wc]> {
+        let c = completions as *mut [ffi::ibv_wc];
+
+        loop {
+            let completions = self.poll(unsafe { &mut *c })?;
+            if !completions.is_empty() {
+                return Ok(completions);
+            }
+
+            let ctx: *mut ffi::ibv_context = unsafe { *self.cq }.context;
+            let errno = unsafe {
+                let ops = &mut { &mut *ctx }.ops;
+                ops.req_notify_cq.as_mut().unwrap()(self.cq, 0)
+            };
+            if errno != 0 {
+                return Err(io::Error::from_raw_os_error(errno));
+            }
+
+            let completions = self.poll(unsafe { &mut *c })?;
```

Comment:
Is another call to `poll` actually warranted here? As in, does the `req_notify_cq` make it likely that there may no be available completions even though there weren't before the call?